File: next_permutation/NextPermutation.java
public class NextPermutation {
    public void nextPermutation(int[] nums) {
        if (nums == null || nums.length < 2) {
            return;
        }
        
        int ptr = nums.length - 2;
        while (ptr >= 0 && nums[ptr + 1] <= nums[ptr]) {
            ptr--;
        }
        if (ptr == -1) {
            reverseArray(nums, 0, nums.length - 1);
            return;
        }
        
        int pivot = nums[ptr];
        int idx = findPivot(nums, ptr, pivot);
        swap(nums, ptr, idx);
        reverseArray(nums, ptr + 1, nums.length - 1);
    }
    
    private void reverseArray(int[] nums, int left, int right) {
        while (left <= right) {
            swap(nums, left++, right--);
        }
    }
    
    private void swap(int[] nums, int left, int right) {
        int tmp = nums[left];
        nums[left] = nums[right];
        nums[right] = tmp;
    }
    
    private int findPivot(int[] nums, int start, int pivot) {
        int ptr = nums.length - 1;
        while (ptr >= start) {
            if (nums[ptr] > pivot) {
                return ptr;
            }
            ptr--;
        }
        return -1;
    }
}

File: next_permutation/Next_Permutation.cpp
class Next_Permutation {
public:
    void nextPermutation(vector<int>& nums) {
        if (nums.empty()) {
            return;
        }
        
        int ptr = nums.size() - 2;
        while (ptr >= 0 && nums[ptr] >= nums[ptr + 1]) {
            ptr--;
        }
        if (ptr == -1) {
            swap_all(nums, 0, nums.size() - 1);
            return;
        }

        swap(nums, ptr, find_target(nums, ptr + 1, nums.size() - 1, nums[ptr]));   
        swap_all(nums, ptr + 1, nums.size() - 1);
    }
    
private:
    int find_target(vector<int>& nums, int left, int right, int target) {
        if (left == right) {
            return left;
        }
        
        while (left < right - 1) {
            int mid = left + (right - left) / 2;
            if (nums[mid] <= target) {
                right = mid;
                continue;
            }
            left = mid;
        }
        
        return nums[right] > target ? right : left;
    }
    
    void swap_all(vector<int>& nums, int left, int right) {
        while (left <= right) {
            swap(nums, left++, right--);
        }
    }
    
    void swap(vector<int>& nums, int left, int right) {
        int tmp = nums[left];
        nums[left] = nums[right];
        nums[right] = tmp;
    }
};